### ISSUE TITLE
Test suite memory errors

### ISSUE THREAD

**Issue by daneos** [sardana-org/sardana#1067]
This is an issue found during manual tests of Jan19 release on CentOS 7 (#1057).

The test suite fails with many `Cannot allocate memory` errors, despite having ~3GB of free RAM.
I tried to isolate the problem with a smaller subset of tests, but it seems it only shows up when the full suite is run.
You can find the full output [here](https://gist.github.com/daneos/c04ca418d5567a17b06d2713cdf86bab).

It doesn't look like a problem with Sardana, more like with the test framework.

@amilan, could you check if the problem is also present on your CentOS environment?

**Comment by reszelaz** [comment]
We saw memory problems with testsuite on Windows and openSUSE 12.1. On Windows we suspect that this may have to do with problems with killing the Tango device servers launched for the needs of tests. Basically this sub-processes does not die. This makes the testsuite process grow in memory (we see that in Task Manager on Windows). But with your observation our assumption may be wrong...

On Debian 9 there are no problems: nothing reported in #1056, travis works on Debian 9 and these kinf of problems never happened before, and also I run the testsuite regularrly on my development machine.

On Debian 9, the standard python is:
```
Python 2.7.13 (default, Sep 26 2018, 18:42:22) 
[GCC 6.3.0 20170516] on linux2
```
I saw from your test output that you must have some _unittest_ residual device servers still defined in the database. You can eliminate them with these loops:

```python
import tango
db = tango.Database()

for s in db.get_server_list("MacroServer/unittest*"):
    db.delete_server(s)

for s in db.get_server_list("Pool/unittest*"):
    db.delete_server(s)
```

Actually I think we should add this to the sardanatestsuite preparation phase.

Could you try to rerun the testsuite after this cleanup. Also ensure that the [sar demo test environment](https://sardana-controls.org/devel/howto_test/test_run_commands.html#sar-demo-test-environment) is running correctly (from the output I see that it was already..).

**Comment by daneos** [comment]
I repeated the tests after cleaning up all old device servers. The result is the same.
Maximum memory utilization during test suite run is ~1.13GB, while the VM has a total of 4GB.
The sar_demo environment is set up correctly.
Default python on CentOS 7 is
```
Python 2.7.5 (default, Oct 30 2018, 23:45:53) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-36)] on linux2
```
but I'm not sure if that can be the issue...

I'm also including other information I posted in #1057 but forgot to add here:
> Disabling memory overcommit protection (`echo 1 > /proc/sys/vm/overcommit_memory`) fixes the issue.

>Enabling swap also fixes the issue, during testsuite run about 260KB of swap is used.

**Comment by reszelaz** [comment]
I was able to reproduce the problem on Debian 9 by disabling swap.

I also confirm that this does not happen with Sardana `2.5.0`. I just re-run the testsuite with the swap off.
The difference is that in that version we were starting only 14 times device servers. Now we start them 40 times.

In your and mine case, the first 25 starts of the device servers work, but the subsequent 15 give the memory error on fork.

**Comment by reszelaz** [comment]
Just to avoid duplication of work..
I will try to reproduce the problem in some simplified example. For the moment I was running dummy testssuite with 50 tests which in the setUp/tearDown start and stop MacroServers. This however does not reproduce the problem. I continue...

**Comment by reszelaz** [comment]
The memory errors were introduced in 23895092 which is a quite big merge commit with a big part of SEP18. I will now try to analyze the changed codebase and see if I'm able to identify the possible reason of these problems. I will keep you updated.

**Comment by daneos** [comment]
Great, thanks! I played around with `ulimit` and `sysctl` to see if I can find why kernel is blocking the allocation, but without much success...
It seems only `overcommit_memory` has any impact on this issue, but there should be no need for overcommit when most of the memory is free.

**Comment by reszelaz** [comment]
For me simply removing the `sardana.pool.test.test_measurementgroup` module avoids this problems, could you also try that?

In this module there are CPU intensive tests. Now I try to selectivelly comment these tests to see if there  is some particular test which breaks everything, or basically a accumulation of the high CPU usage. Maybe high CPU affects the garbage collection? Actually we saw on the openSUSE12.1 a high memory footprint of the testsuite

**Comment by reszelaz** [comment]
For the moment I have no clue on these tests yet. But it must have something to do with the free memory. The same set of tests fails when I'm close to the memory limit, and when I close the PyCharm and relese like 2Gb everything is fine.

**Comment by daneos** [comment]
> For me simply removing the `sardana.pool.test.test_measurementgroup` module avoids this problems, could you also try that?

Yes, it works. But running only tests in `sardana.pool.test.test_measurementgroup` without the rest of test suite doesn't fail either.

**Comment by reszelaz** [comment]
> But running only tests in `sardana.pool.test.test_measurementgroup` without the rest of test suite doesn't fail either.

I think the problem is when these tests are executed from the same python process together with tests which execute `os.fork()` for spawning device servers. Separatelly there are no problems.

I discovered that recently we changed one test base class and create there 20 dummy 2D elements and each of these internally create a 1024 x 1024 numpy array of zeros! The process virtual memory grows while the said tests are run. These tests are CPU intensive and I'm not sure if we don't have circular references somewhere... so garbage collection may be tricky here. Anyway, I don't have a in mind a clear explanation for that and simply propose to not generate 20 dummy 2D elements but only one - see #1071. At the end we don't need so many... Maybe some python/OS expert could comment and provide a more technical explanation of the root cause?

We tried the said fix with openSUSE 12.1, Windows and the memory errors disappeared.
